datalake/src/main/java/com/sequenceiq/datalake/controller/DatalakeEventController.java

```diff
@@ -0,0 +1,136 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.StreamingOutput;
+
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Controller;
+
+import com.sequenceiq.authorization.annotation.CustomPermissionCheck;
+import com.sequenceiq.authorization.utils.EventAuthorizationDto;
+import com.sequenceiq.authorization.utils.EventAuthorizationUtils;
+import com.sequenceiq.cloudbreak.auth.crn.Crn;
+import com.sequenceiq.cloudbreak.auth.crn.CrnParseException;
+import com.sequenceiq.cloudbreak.common.exception.BadRequestException;
+import com.sequenceiq.cloudbreak.common.json.JsonUtil;
+import com.sequenceiq.cloudbreak.structuredevent.event.StructuredEventType;
+import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPStructuredEvent;
+import com.sequenceiq.cloudbreak.structuredevent.service.db.CDPStructuredEventDBService;
+import com.sequenceiq.datalake.entity.SdxCluster;
+import com.sequenceiq.datalake.repository.SdxClusterRepository;
+import com.sequenceiq.sdx.api.endpoint.DatalakeEventEndpoint;
+
+@Controller
+public class DatalakeEventController implements DatalakeEventEndpoint {
+
+    @Inject
+    private CDPStructuredEventDBService cdpStructuredEventDBService;
+
+    @Inject
+    private EventAuthorizationUtils eventAuthorizationUtils;
+
+    @Inject
+    private SdxClusterRepository sdxClusterRepository;
+
+    /**
+     * Retrieves audit events for all datalakes that have been associated with the provided environment CRN.
+     *
+     * @param resourceCrn an environment CRN
+     * @param types       types of structured events to retrieve
+     * @param page        the page index
+     * @param size        the size of the page to retrieve
+     * @return structured events for all datalakes related to the resource CRN of an environment
+     */
+    @Override
+    @CustomPermissionCheck
+    public List<CDPStructuredEvent> getAuditEvents(String resourceCrn, List<StructuredEventType> types, Integer page, Integer size) {
+        PageRequest pageable = PageRequest.of(page, size, Sort.by("timestamp").descending());
+        // datalake crns related to the provided environment CRN.
+        List<String> datalakeCrns = getRelatedDatalakeCrns(resourceCrn);
+
+        // actually retrieve events from DB
+        List<CDPStructuredEvent> events = cdpStructuredEventDBService.getPagedEventsOfResources(types, datalakeCrns, pageable).getContent();
+        if (events.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        eventAuthorizationUtils.checkPermissionBasedOnResourceTypeAndCrn(collectDtosFromEvents(events));
+
+        return events;
+    }
+
+    private String getAccountId(String crnString) {
```

Comment:
I would put `@ValidCrn` annotation on the `getAuditEvents` `resourceCrn` parameter and just use `Crn.safeFromString(resourceCrn).getAccountId()`, in this case we would not need a separate method

Reply:
Done, thanks for the tip!